File: cpp/leetcode/030.cpp
class Solution {
  public:
    vector<int> findSubstring(string s, vector<string> &words) {
        vector<int> res;
        unordered_map<string, int> wmap;
        int m = words.size(), n = words[0].size();
        for (string str : words) {
            wmap[str]++;
        }
        for (int i = 0; i < s.size() - m * n + 1; i++) {
            int j = 0;
            unordered_map<string, int> smap;
            for (; j < m; j++) {
                string key = s.substr(i + j * n, n);
                if (wmap.find(key) == wmap.end()) {
                    break;
                }
                if (++smap[key] > wmap[key]) {
                    break;
                }
            }
            if (j == m) {
                res.push_back(i);
            }
        }
        return res;
    }
};

File: python/leetcode/030.py
class Solution(object):
    def findSubstring(self, s, words):
        n, m = len(words), len(words[0])  # m word length
        cnts, res = {}, []   # total words length
        for word in words:
            cnts[word] = cnts.get(word, 0) + 1
        for i in range(len(s) - m * n + 1):
            j, cur = 0, {}
            while j < n:
                word = s[i + j * m: i + (j + 1) * m]
                if word not in cnts:
                    break
                cur[word] = cur.get(word, 0) + 1
                if cur[word] > cnts[word]:
                    break
                j += 1
            if j == n:
                res.append(i)
        return res
